keylime/cloud_agent.py

```diff
@@ -158,90 +160,90 @@ def do_GET(self):
             return
-        
+
         post_body = self.rfile.read(content_length)
-        json_body = json.loads(post_body)
-            
-        b64_encrypted_key = json_body['encrypted_key']
+        yaml_body = yaml.safe_load(post_body)
+
+        b64_encrypted_key = yaml_body['encrypted_key']
         decrypted_key = crypto.rsa_decrypt(self.server.rsaprivatekey,base64.b64decode(b64_encrypted_key))
-        
+
         have_derived_key = False
 
         if rest_params["keys"] == "ukey":
             self.server.add_U(decrypted_key)
-            self.server.auth_tag = json_body['auth_tag']
-            self.server.payload = json_body.get('payload',None)
-            
+            self.server.auth_tag = yaml_body['auth_tag']
+            self.server.payload = yaml_body.get('payload',None)
             have_derived_key = self.server.attempt_decryption(self)
         elif rest_params["keys"] == "vkey":
             self.server.add_V(decrypted_key)
             have_derived_key = self.server.attempt_decryption(self)
         else:
             logger.warning('POST returning  response. uri not supported: ' + self.path)
-            common.echo_json_response(self, 400, "uri not supported")
+            common.echo_yaml_response(self, 400, "uri not supported")
             return
         logger.info('POST of %s key returning 200'%(('V','U')[rest_params["keys"] == "ukey"]))
-        common.echo_json_response(self, 200, "Success")
-        
+        common.echo_yaml_response(self, 200, "Success")
+
         # no key yet, then we're done
         if not have_derived_key:
             return
-        
-        # woo hoo we have a key 
+
+        # woo hoo we have a key
         # ok lets write out the key now
         secdir = secure_mount.mount() # confirm that storage is still securely mounted
-        
+
         # clean out the secure dir of any previous info before we extract files
         if os.path.isdir("%s/unzipped"%secdir):
             shutil.rmtree("%s/unzipped"%secdir)
-        
+
         # write out key file
         f = open(secdir+"/"+self.server.enc_keyname,'w')
-        f.write(base64.b64encode(self.server.K))
+        f.write(base64.b64encode(self.server.K).decode())
         f.close()
-        
+
         #stow the U value for later
         tpm.write_key_nvram(self.server.final_U)
-        
+
         # optionally extend a hash of they key and payload into specified PCR
         tomeasure = self.server.K
-        
+
         # if we have a good key, now attempt to write out the encrypted payload
         dec_path = "%s/%s"%(secdir, config.get('cloud_agent',"dec_payload_file"))
         enc_path = "%s/encrypted_payload"%common.WORK_DIR
-        
+
         dec_payload = None
         enc_payload = None
-        
+
         if self.server.payload is not None:
-            dec_payload = crypto.decrypt(self.server.payload, str(self.server.K))
+            dec_payload = crypto.decrypt(self.server.payload, self.server.K)
```

Comment:
Make sure the `elif` block below matches how we're doing it here.  Right now, if we're using a previous payload then we load with `dec_payload = crypto.decrypt(enc_payload,str(self.server.K))` (note the conversion to _str_ still)

Reply:
This is what we get with `dec_payload = crypto.decrypt(self.server.payload, str(self.server.K))`


```
2019-07-31 12:02:49.703 - keylime.cloudagent - INFO - POST of V key returning 200
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 58382)
Traceback (most recent call last):
  File "/usr/lib64/python3.7/socketserver.py", line 650, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/lib64/python3.7/socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib64/python3.7/socketserver.py", line 720, in __init__
    self.handle()
  File "/usr/lib64/python3.7/http/server.py", line 426, in handle
    self.handle_one_request()
  File "/usr/lib64/python3.7/http/server.py", line 414, in handle_one_request
    method()
  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/cloud_agent.py", line 246, in do_POST
    dec_payload = crypto.decrypt(self.server.payload, str(self.server.K))
  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/crypto.py", line 157, in decrypt
    cipher = AES.new(key, AES.MODE_GCM, nonce = nonce)
  File "/usr/local/lib64/python3.7/site-packages/Cryptodome/Cipher/AES.py", line 232, in new
    return _create_cipher(sys.modules[__name__], key, mode, *args, **kwargs)
  File "/usr/local/lib64/python3.7/site-packages/Cryptodome/Cipher/__init__.py", line 79, in _create_cipher
    return modes[mode](factory, **kwargs)
  File "/usr/local/lib64/python3.7/site-packages/Cryptodome/Cipher/_mode_gcm.py", line 622, in _create_gcm_cipher
    return GcmMode(factory, key, nonce, mac_len, kwargs, ghash_c)
  File "/usr/local/lib64/python3.7/site-packages/Cryptodome/Cipher/_mode_gcm.py", line 225, in __init__
    **cipher_params
  File "/usr/local/lib64/python3.7/site-packages/Cryptodome/Cipher/AES.py", line 232, in new
    return _create_cipher(sys.modules[__name__], key, mode, *args, **kwargs)
  File "/usr/local/lib64/python3.7/site-packages/Cryptodome/Cipher/__init__.py", line 79, in _create_cipher
    return modes[mode](factory, **kwargs)
  File "/usr/local/lib64/python3.7/site-packages/Cryptodome/Cipher/_mode_ecb.py", line 215, in _create_ecb_cipher
    cipher_state = factory._create_base_cipher(kwargs)
  File "/usr/local/lib64/python3.7/site-packages/Cryptodome/Cipher/AES.py", line 93, in _create_base_cipher
    raise ValueError("Incorrect AES key length (%d bytes)" % len(key))
ValueError: Incorrect AES key length (102 bytes)
```

Reply:
and debug output for candidate and decrypted_U

```
decrypted_U: b'\x9eM:!\xca\n\xca\xabn"\xc3f\x11f\xb2:5\x9e\x1a<\xf6p\xae\xef\xb9[\xaaE\xd3W\xfb\x13'
candidate_key: bytearray(b'\xdf\x1c(\x0c\xc5\xbd8\x08\xd2\xc5\xc6\x8em\xf3\xde0Y\xbf\n\x15:\xed\x1f.V\x98\x1d\xc9\xb5:\x9f\xf7')
```

Reply:
I noticed in the latest commit you added (https://github.com/keylime/keylime/pull/126/commits/436ce3c73f7cc63f8fc2a45ab7fa2da91da40f56#diff-242fc79fc17586955d7b4bcd2c12f4d5R637): 

````python
    async def do_verify(self):
        """initiaite v, agent_id and ip
        """initiaite v, agent_id and ip
        initiate the cloudinit sequence"""
        initiate the cloudinit sequence"""
        challenge = TPM_Utilities.random_password(20)
        challenge = TPM_Utilities.random_password(20)
        numtries = 0
        encoded = base64.b64encode(self.K).decode('utf-8')
````

This might be causing the issue you're seeing, since the crypto libraries are expecting a raw key `self.K` and not base64-encoded. 

Does `encoded = base64.b64encode(self.K).decode('utf-8')` need to be removed?

Reply:
Note that I think the way you've fixed the `if` block there is correct, but just don't forget to update the `elif` block as well.

From: 
https://github.com/keylime/keylime/blob/436ce3c73f7cc63f8fc2a45ab7fa2da91da40f56/keylime/cloud_agent.py#L245-L253

To: 
````python
        if self.server.payload is not None:
            dec_payload = crypto.decrypt(self.server.payload, self.server.K)
            enc_payload = self.server.payload
        elif os.path.exists(enc_path):
            # if no payload provided, try to decrypt one from a previous run stored in encrypted_payload
            with open(enc_path,'r') as f:
                enc_payload = f.read()
            try:
                dec_payload = crypto.decrypt(enc_payload,self.server.K)
````

Reply:
I have removed that for now, still getting a mismatch:

```
mac: 129ce651fabc4942b59a08efb7eb0a1444ddd04909e92e01cac7fce8d31ed9cb91edfeb843e7d9a61c4540622b7a08bf
mac: <class 'str'>
ex_mac: 8b63b4df48a59e9928d2c2f004a0673fd85c67c3692398ab504a3fd8563a58133d734ac4df51417f13f1960ad129047e
ex_mac: <class 'str'>
```

Reply:
This seems to be working on my side, but I get hit with a random `Error: HTTP 599: Timeout while connecting` error on the verifier on occasion.  

I do get an issue on dealing with the payload for the cloud_agent though, related to appending byte arrays to each other.

Reply:
I have made some progress here:

```
2019-07-31 13:26:25.806 - keylime.tenant - ERROR - Response code 400: Bootstrap key not yet available.
2019-07-31 13:26:25.806 - keylime.tenant - WARNING - Key derivation not yet complete...trying again in 1.0 seconds...Ctrl-C to stop
2019-07-31 13:26:26.818 - keylime.tenant - INFO - Key derivation successful
```

However this couples with:

```
  File "/usr/local/lib/python3.7/site-packages/keylime-1.2-py3.7.egg/keylime/cloud_agent.py", line 264, in do_POST
    f.write(self.server.payload)
TypeError: write() argument must be str, not bytes
```

Yet if I encode self.server.payload or add a `b` to 

```
with open(enc_path,'wb') as f:
                f.write(self.server.payload)
```

Key derivation fails. I need to work out end to end how to handle the payloads.

Reply:
Should I push my new code @jetwhiz ?

Reply:
Your `436ce3c` commit seems to be working for me.  What's changed in your new code?  Just adding that binary flag to the file write for payload? 

Did you also add the `b` to the corresponding read part?  

https://github.com/keylime/keylime/blob/436ce3c73f7cc63f8fc2a45ab7fa2da91da40f56/keylime/cloud_agent.py#L250-L251

Reply:
I will get you the whole file, its working completely now:

https://gist.github.com/lukehinds/ecf5fe0cceb366bbd771cf6b2e7b2c78

Reply:
Nice!  Go ahead and push that commit and I'll take a look